### PR TITLE
Also write a YAML file for the final map.

### PR DESCRIPTION
**Pull request by wohe**
This allows the saved map to be served by map_server.

### CODE REVIEW

File: cartographer_ros/src/map_writer.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "glog/logging.h"
+#include "yaml-cpp/yaml.h"
+
+namespace cartographer_ros {
+
+namespace {
+
+void WriteOccupancyGridToPgm(const ::nav_msgs::OccupancyGrid& grid,
+                             const std::string& filename) {
+  LOG(INFO) << "Saving final map to '" << filename << "'...";
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: This function does not know if this is the 'final map'. Hoist comment to the callsite?

File: cartographer_ros/src/map_writer.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  std::ofstream pgm_file(filename, std::ios::out | std::ios::binary);
+  const std::string header = "P5\n# Cartographer map; " +
+                             std::to_string(grid.info.resolution) +
+                             " m/pixel\n" + std::to_string(grid.info.width) +
+                             " " + std::to_string(grid.info.height) + "\n255\n";
+  pgm_file.write(header.data(), header.size());
+  for (size_t y = 0; y < grid.info.height; ++y) {
+    for (size_t x = 0; x < grid.info.width; ++x) {
+      size_t i = x + (grid.info.height - y - 1) * grid.info.width;
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: const

File: cartographer_ros/src/map_writer.cc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      if (grid.data[i] >= 0 && grid.data[i] <= 100) {
+        pgm_file.put((100 - grid.data[i]) * 255 / 100);
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: should this be float math? if so, add dots after the numbers.

File: cartographer_ros/src/map_writer.cc [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      } else {
+        constexpr uint8_t kUnknownValue = 128;
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: add a comment why we choose 128 for unknown

File: cartographer_ros/src/map_writer.cc [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+  pgm_file.close();
+  CHECK(pgm_file) << "Writing '" << filename << "' failed.";
+}
+
+void WriteOccupancyGridInfoToYaml(const ::nav_msgs::OccupancyGrid& grid,
+                                  const std::string& map_filename,
+                                  const std::string& yaml_filename) {
+  LOG(INFO) << "Saving final map info to '" << yaml_filename << "'...";
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: hoist to callsite?

File: cartographer_ros/src/map_writer.cc [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  std::ofstream yaml_file(yaml_filename, std::ios::out | std::ios::binary);
+  {
+    YAML::Emitter out(yaml_file);
+    out << YAML::BeginMap;
+    // TODO(whess): Use basename only?
+    out << YAML::Key << "image" << YAML::Value << map_filename;
+    out << YAML::Key << "resolution" << YAML::Value << grid.info.resolution;
+    constexpr double kYawButMaybeIgnored = 0.;
```
Comment (SirVer) [inline cartographer_ros/src/map_writer.cc]: Add a comment about the dubious documentation?

**Comment by wohe** [comment]
PTAL

**Pull request merged**